# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Nursery-inl.h
# Commit: 902e72cc89ff
# Full Hash: 902e72cc89ff77342b4c7ed5c1afb3e18bc20116
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2021-01-21 15:33:26
# Description:
#   Bug 1687690 - Relax assertion now that tenured heap objects can live at the end of chunks r=sfink
#   
#   I'm not 100% sure that this is the problem but it seems plausible.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D102424
# ==============================================================================

diff -r f1e52c54a12e -r 902e72cc89ff js/src/gc/Nursery-inl.h
--- a/js/src/gc/Nursery-inl.h	Thu Jan 21 14:11:55 2021 +0200
+++ b/js/src/gc/Nursery-inl.h	Thu Jan 21 10:38:55 2021 +0000
@@ -81,13 +81,10 @@
 
 inline void js::Nursery::setDirectForwardingPointer(void* oldData,
                                                     void* newData) {
-  MOZ_ASSERT(isInside(oldData));
-
-  // Bug 1196210: If a zero-capacity header lands in the last 2 words of a
-  // jemalloc chunk abutting the start of a nursery chunk, the (invalid)
-  // newData pointer will appear to be "inside" the nursery.
-  MOZ_ASSERT(!isInside(newData) ||
-             (uintptr_t(newData) & js::gc::ChunkMask) == 0);
+  // If a zero-capacity elements header lands right at the end of a chunk then
+  // elements data will appear to be in the next chunk.
+  MOZ_ASSERT(isInside(oldData) || (uintptr_t(oldData) & gc::ChunkMask) == 0);
+  MOZ_ASSERT(!isInside(newData) || (uintptr_t(newData) & gc::ChunkMask) == 0);
 
   new (oldData) BufferRelocationOverlay{newData};
 }